render-wasm/docs/tile_rendering.md

```diff
@@ -0,0 +1,202 @@
+  - Allow UI updates
+  - Handle user input
+  - Maintain consistent frame rates
+
+> 💡 This technique is similar to how web mapping tools or graphics engines progressively load and render content.
+
+This incremental rendering strategy ensures:
+- The most relevant (visible) tiles appear first.
+- Remaining tiles are rendered progressively, minimizing perceived latency.
+- The user interface remains responsive during rendering operations.
+
+### Interaction and Updates
+
+When a user interacts (creates, moves, or edits a shape):
+
+- Affected tiles are computed using the shape's bounding box.
+- `TileHashMap` updates to reflect tile-shape relationships.
+- Affected tiles are invalidated in the cache.
+- Only those tiles are re-rendered.
+
+When zooming or panning:
+
+- New tiles entering the `visible_rect` are identified.
+- These tiles are either:
+  - Pulled from the cache, or  
+  - Rendered fresh if not already available.
+
+### Handling Zoom and Scale
+
+- Tile resolution depends on zoom level (`scale`).
+- When zoomed in significantly, higher-resolution tiles are required.
+- If the zoom changes beyond a threshold, the cache may be invalidated, and tiles re-rendered at the new scale.
+
+## Why cache tiles?
+
+Tile caching (via `TileTextureCache`) boosts performance:
+
+- Prevents unnecessary rendering.
+- Only a limited number of tiles are kept in memory (`TEXTURES_CACHE_CAPACITY`).
+- Distant tiles are periodically evicted using an LRU-like strategy (based on visibility).
+
+## Scaling and Tile Size Utilities
+
+Tile size in screen space is constant, but visually it depends on zoom level (`scale`).
+
+To deal with this, the system uses utility functions:
+
+- `get_tile_size(scale)`: Gets the logical size of a tile at the current scale.
+- `get_tile_rect(tile, scale)`: Gets the rectangular area a tile covers.
+- `get_tile_pos(tile, scale)`: Gets the tile’s position in canvas coordinates.
+
+These allow rendering logic to adapt tile positions/sizes dynamically.
+
+## Selective Re-rendering
+
+The system enables:
+
+- Efficient updates when the viewport changes.
+- Targeted re-rendering when a shape changes (only the affected tiles).
+- Reduced GPU/CPU work on complex documents.
+
+## Rendering Pipeline Summary (Per Frame)
+
+1. **Input**: `document_state` + `Viewbox`  
+   - Shapes, layer order, transforms + visible area.
+
+2. **Compute `TileViewbox`**  
+   - From `Viewbox`, generate visible + interest regions.
+
+3. **Determine Tiles to Render**  
+   - Compare current tiles vs cached ones.
+   - Detect cache hits and misses.
+   - Invalidate stale tiles if viewport changed significantly.
+
+4. **Render From Cache**  
+   - For cached tiles:
+     - Fetch image from `TileTextureCache`.
+     - Blit to `Surface::Target`.
+
+5. **Render New Tiles**  
+   - For non-cached tiles:
+     - Find intersecting shapes (`TileHashMap`).
+     - Render layers in this order:
+       - Fills → Strokes → Shadows → Composite → Target
```

Comment:
We already mention the ordering above, and I don't think it's that important to get repeated here. We can combine these points with the ones below:

- Render the tile layers on their `skia::Surface` instances, using blending, opacity, masking, etc. when needed.